server/server.go

```diff
@@ -0,0 +1,62 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/medibloc/panacea-doracle/config"
+	"github.com/medibloc/panacea-doracle/server/service"
+	"net/http"
+	"os"
+	"os/signal"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func Run(conf *config.Config) error {
+	svc, err := service.New(conf)
+	if err != nil {
+		return fmt.Errorf("failed to create service: %w", err)
+	}
+	defer svc.Close()
+
+	server := &http.Server{
+		WriteTimeout: 15 * time.Second,
+		ReadTimeout:  15 * time.Second,
+	}
+
+	httpServerErrCh := make(chan error, 1)
+	go func() {
+		log.Infof("Decentralized Oracle Started")
+		if err := server.ListenAndServe(); err != nil {
+			if !errors.Is(err, http.ErrServerClosed) {
+				httpServerErrCh <- err
+			} else {
+				close(httpServerErrCh)
+			}
+		}
+	}()
+
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, os.Interrupt)
+	select {
+	case err := <-httpServerErrCh:
+		if err != nil {
+			log.Errorf("http server was closed with an error: %v", err)
+		}
+	case <-signalCh:
+		log.Info("signal detected")
+	}
+	log.Info("starting the graceful shutdown")
+
+	log.Info("terminating HTTP server")
+	ctxTimeout, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	if err := server.Shutdown(ctxTimeout); err != nil {
+		return fmt.Errorf("error occurs while server shutting down: %w", err)
+	}
+
+	return nil
```

Comment:
Good to have this HTTP server as you said in https://github.com/medibloc/panacea-doracle/pull/3/files#r929439137.
Then, please keep in mind that this code block needs to be extracted to the `main()` function in the future, because we are going to run one more goroutine for subscribing Tendermint events. Then, after getting a OS signal (SIGINT, or so), you need to shutdown the subscriber as well as the HTTP server.

Reply:
Yes. I remember we made an issue related refactoring graceful shutdown before (https://github.com/medibloc/panacea-coracle/issues/34). I will move this graceful shutdown to `main()` in the following PR, because I think it also needs to be done in code structuring. Thank you for reminding me!